src/services/cache-service.ts

```diff
@@ -2,63 +2,109 @@
  * Caches the response of any promise value of a loading function and prevents duplicate calls
  * @author Mike Reiche <mike.reiche@t-systems.com>
  */
+import {TimeComponents, toMilliseconds} from "../utils/time";
+
 export class CacheService {
-    private _cacheContainer = {};
-    private _defaultCacheTtlSeconds = 10;
+    private _cacheContainer:{[key: string]: [number, Promise<any>]} = {};
+    private _defaultCacheTtl:TimeComponents = {seconds: 10};
 
+    /**
+     *
+     * @param key The key for cached value
+     * @param loadingFunction The loading function returning a promise of the value.
+     * @param ttl The cache timeout for the response.
+     */
     getForKeyWithLoadingFunction <T> (
         key:string,
         loadingFunction:() => Promise<T>,
-        cacheTtlSeconds = this._defaultCacheTtlSeconds
+        ttl:number|TimeComponents = this._defaultCacheTtl
     ):Promise<T> {
-        let cacheEntry = this._cacheContainer[key];
-        const now = (Date.now()/1000);
-        if (cacheEntry !== undefined) {
-            if (cacheEntry[0] >= now - cacheTtlSeconds) {
-                return cacheEntry[1];
-            }
+        const cacheEntry = this._cacheContainer[key];
+        const now = Date.now();
+
+        if (cacheEntry !== undefined && cacheEntry[0] >= now) {
+            return cacheEntry[1];
         }
 
+        ttl = CacheService.wrapTtl(ttl);
+        const cacheTtlMs = toMilliseconds(ttl);
+        const validUntil = now + cacheTtlMs;
         this._cacheContainer[key] = [
-            now,
-            loadingFunction().then(object=>{
-                return this._cacheContainer[key][1] = Promise.resolve(object);
+            validUntil,
+            loadingFunction().then(response => {
+                const cachedResponse = Promise.resolve(response);
+                this._cacheContainer[key] = [validUntil, cachedResponse];
```

Comment:
I changed the logic from "time to live" to "valid until". So we have always the fully UTC timestamp until which the cache entry is valid.

- The cache entry is getting immediately alive when requested.
- Passing a loading function is like defining a "fallback" when no valid cache entry exists, it will then become the new cache entry.

Reply:
Ok, "valid until" is a different thing and makes sense to me in this context.